lib/src/ignore.dart

```diff
@@ -0,0 +1,415 @@
+          index != -1;
+          index = path.lastIndexOf('/', index - 1)) {
+        final result = testRulesForPrefix(path.substring(0, index));
+        if (result != null) return result;
+      }
+      return testRulesForPrefix('') == true;
+    }
+
+    for (var slashIndex = path.indexOf('/');
+        slashIndex != -1;
+        slashIndex = path.indexOf('/', slashIndex + 1)) {
+      if (ignoresSubpath(path.substring(0, slashIndex + 1))) {
+        // If a folder above [path] is ignored the pattern cannot be un-ignored.
+        return true;
+      }
+    }
+    return ignoresSubpath(path);
+  }
+}
+
+class GitIgnoreParseResult {
+  // The parsed pattern.
+  final String pattern;
+
+  // The resulting matching rule. `null` if the pattern was empty or invalid.
+  final GitIgnoreRule rule;
+
+  // An invalid pattern is also considered empty.
+  bool get empty => rule == null;
+  bool get valid => exception == null;
+
+  // For invalid patterns this contains a description of the problem.
+  final FormatException exception;
+
+  GitIgnoreParseResult(this.pattern, this.rule) : exception = null;
+  GitIgnoreParseResult.invalid(this.pattern, this.exception) : rule = null;
+  GitIgnoreParseResult.empty(this.pattern)
+      : rule = null,
+        exception = null;
+}
+
+class GitIgnoreRule {
+  /// A regular expression that represents this rule.
+  final RegExp pattern;
+  final bool negative;
+
+  /// The String this pattern was generated from.
+  final String original;
+
+  GitIgnoreRule(this.pattern, this.negative, this.original);
+
+  @override
+  String toString() {
+    // TODO: implement toString
+    return '$original -> $pattern';
+  }
+}
+
+/// [onInvalidPattern] can be used to handle parse failures. If
+/// [onInvalidPattern] is `null` invalid patterns are ignored.
+Map<String, List<GitIgnoreRule>> parseIgnorePatterns(
+    Map<String, Iterable<String>> patternsHierarchy, bool ignoreCase,
+    {void Function(String pattern, FormatException exception)
+        onInvalidPattern}) {
+  ArgumentError.checkNotNull(patternsHierarchy, 'patterns');
+  ArgumentError.checkNotNull(ignoreCase, 'ignoreCase');
+  if (patternsHierarchy.values.contains(null)) {
+    throw ArgumentError.value(
+        patternsHierarchy, 'patterns', 'may not contain null');
+  }
+
+  return patternsHierarchy.map((directory, patterns) {
+    final parseResults = patterns
+        .map((s) => s.split('\n'))
+        .expand((e) => e)
+        .map((pattern) => parseIgnorePattern(pattern, ignoreCase));
+    if (onInvalidPattern != null) {
+      for (final invalidResult
+          in parseResults.where((result) => !result.valid)) {
+        onInvalidPattern(invalidResult.pattern, invalidResult.exception);
+      }
+    }
+    return MapEntry(directory,
+        parseResults.where((r) => !r.empty).map((r) => r.rule).toList());
+  });
+}
+
+GitIgnoreParseResult parseIgnorePattern(String pattern, bool ignoreCase,
+    {source}) {
```

Comment:
Should we add some typing to `source` or pass it along when we call from `parseIgnorePatterns`.

Reply:
I removed source. It can just be the pattern itself.

Reply:
Removed the source. It can just be the pattern itself...